File: test_aes_compatibility.py
#!/usr/bin/env python3
"""
AES Compatibility Test - Client/Server Validation

This script validates that our client's AES-256-CBC implementation produces
output that is compatible with the server's decryption process.

Key Requirements:
- AES-256-CBC mode
- 32-byte (256-bit) key
- Zero IV (16 bytes of 0x00)  
- PKCS7 padding
- Client encrypts, server decrypts

The goal is to identify any format incompatibilities before testing with live server.
"""

import os
import sys
import binascii
from typing import Tuple, Optional

# Import the server's crypto compatibility layer
sys.path.append(os.path.join(os.path.dirname(__file__), 'server'))
from crypto_compat import AES, unpad, get_random_bytes

def generate_test_aes_key() -> bytes:
    """Generate a 32-byte AES-256 key for testing."""
    return get_random_bytes(32)

def simulate_client_aes_encryption(data: bytes, aes_key: bytes) -> bytes:
    """
    Simulate the client's AES encryption using the same method as AESWrapper.cpp.
    
    This mirrors the client's encryption:
    - AES-256-CBC mode
    - Zero IV (all zeros)
    - PKCS7 padding
    """
    if len(aes_key) != 32:
        raise ValueError(f"AES key must be 32 bytes, got {len(aes_key)}")
    
    # Create zero IV (16 bytes of 0x00) - matches client's useStaticZeroIV=true
    zero_iv = b'\x00' * 16
    
    # Create AES cipher in CBC mode with zero IV
    cipher = AES.new(aes_key, AES.MODE_CBC, iv=zero_iv)
    
    # Add PKCS7 padding manually (to match Crypto++ behavior)
    block_size = 16
    padding_length = block_size - (len(data) % block_size)
    padded_data = data + bytes([padding_length] * padding_length)
    
    # Encrypt the padded data
    encrypted_data = cipher.encrypt(padded_data)
    
    print(f"[CLIENT-SIM] Encrypted {len(data)} bytes to {len(encrypted_data)} bytes")
    print(f"[CLIENT-SIM] Key: {binascii.hexlify(aes_key).decode()}")
    print(f"[CLIENT-SIM] IV: {binascii.hexlify(zero_iv).decode()}")
    print(f"[CLIENT-SIM] Plaintext: {binascii.hexlify(data).decode()}")
    print(f"[CLIENT-SIM] Padded: {binascii.hexlify(padded_data).decode()}")
    print(f"[CLIENT-SIM] Ciphertext: {binascii.hexlify(encrypted_data).decode()}")
    
    return encrypted_data

def simulate_server_aes_decryption(encrypted_data: bytes, aes_key: bytes) -> bytes:
    """
    Simulate the server's AES decryption using the exact method from server.py.
    
    This mirrors the server's decryption:
    - AES-256-CBC mode  
    - Zero IV (all zeros)
    - PKCS7 padding removal using unpad()
    """
    if len(aes_key) != 32:
        raise ValueError(f"AES key must be 32 bytes, got {len(aes_key)}")
    
    # Create zero IV (16 bytes of 0x00) - matches server's iv=b'\0' * 16
    zero_iv = b'\x00' * 16
    
    # Create AES cipher in CBC mode with zero IV (exact server code)
    cipher_aes = AES.new(aes_key, AES.MODE_CBC, iv=zero_iv)
    
    # Decrypt the data
    decrypted_padded = cipher_aes.decrypt(encrypted_data)
    
    # Remove PKCS7 padding using server's unpad method
    # AES block size is always 16 bytes
    decrypted_data = unpad(decrypted_padded, 16)
    
    print(f"[SERVER-SIM] Decrypted {len(encrypted_data)} bytes to {len(decrypted_data)} bytes")
    print(f"[SERVER-SIM] Key: {binascii.hexlify(aes_key).decode()}")
    print(f"[SERVER-SIM] IV: {binascii.hexlify(zero_iv).decode()}")
    print(f"[SERVER-SIM] Ciphertext: {binascii.hexlify(encrypted_data).decode()}")
    print(f"[SERVER-SIM] Decrypted+Padded: {binascii.hexlify(decrypted_padded).decode()}")
    print(f"[SERVER-SIM] Final: {binascii.hexlify(decrypted_data).decode()}")
    
    return decrypted_data

def test_aes_round_trip(test_data: bytes, aes_key: bytes) -> Tuple[bool, Optional[str]]:
    """
    Test a complete encrypt/decrypt round trip to validate compatibility.
    
    Returns:
        (success, error_message)
    """
    try:
        print(f"\n=== ROUND TRIP TEST ===")
        print(f"Original data: {test_data}")
        print(f"Original hex: {binascii.hexlify(test_data).decode()}")
        
        # Step 1: Encrypt using client method
        encrypted = simulate_client_aes_encryption(test_data, aes_key)
        
        print(f"\nEncryption successful: {len(encrypted)} bytes")
        
        # Step 2: Decrypt using server method  
        decrypted = simulate_server_aes_decryption(encrypted, aes_key)
        
        print(f"Decryption successful: {len(decrypted)} bytes")
        print(f"Decrypted data: {decrypted}")
        
        # Step 3: Compare original vs decrypted
        if test_data == decrypted:
            print("✅ SUCCESS: Data matches after round trip!")
            return True, None
        else:
            error_msg = f"❌ FAILURE: Data mismatch!\nOriginal: {test_data}\nDecrypted: {decrypted}"
            print(error_msg)
            return False, error_msg
            
    except Exception as e:
        error_msg = f"❌ EXCEPTION during round trip: {str(e)}"
        print(error_msg)
        return False, error_msg

def run_compatibility_tests():
    """Run a comprehensive suite of AES compatibility tests."""
    print("=== AES COMPATIBILITY TEST SUITE ===")
    print("Testing client AES encryption vs server AES decryption")
    print()
    
    # Generate test AES key
    aes_key = generate_test_aes_key()
    print(f"Test AES Key (32 bytes): {binascii.hexlify(aes_key).decode()}")
    
    # Test cases with different data sizes and content
    test_cases = [
        # Basic text
        b"Hello, World!",
        
        # Empty data  
        b"",
        
        # Single byte
        b"A",
        
        # Exact block size (16 bytes)
        b"1234567890123456",
        
        # Block size + 1
        b"12345678901234567",
        
        # Multiple blocks  
        b"This is a longer test message that spans multiple AES blocks for thorough testing.",
        
        # Binary data
        b"\x00\x01\x02\x03\xff\xfe\xfd\xfc\xaa\xbb\xcc\xdd\xee\x11\x22\x33",
        
        # Typical file content simulation
        b"This is test file content that would be encrypted and sent to the server for backup.",
        
        # Large-ish data
        b"X" * 1000,
    ]
    
    results = []
    
    for i, test_data in enumerate(test_cases):
        print(f"\n--- Test Case {i+1}: {len(test_data)} bytes ---")
        success, error = test_aes_round_trip(test_data, aes_key)
        results.append((success, len(test_data), error))
    
    # Summary
    print(f"\n=== TEST RESULTS SUMMARY ===")
    passed = sum(1 for success, _, _ in results if success)
    total = len(results)
    
    print(f"Tests passed: {passed}/{total}")
    
    if passed == total:
        print("🎉 ALL TESTS PASSED - Client/Server AES compatibility confirmed!")
        print("\nThe client's AES implementation should be compatible with the server.")
        print("Key implementation details:")
        print("- AES-256-CBC mode")
        print("- 32-byte key")
        print("- Zero IV (16 bytes of 0x00)")
        print("- PKCS7 padding")
    else:
        print("⚠️  SOME TESTS FAILED - Compatibility issues detected!")
        print("\nFailed tests:")
        for i, (success, size, error) in enumerate(results):
            if not success:
                print(f"  Test {i+1} ({size} bytes): {error}")
                
    return passed == total

def demonstrate_exact_server_decryption():
    """
    Show exactly what the client needs to output for successful server decryption.
    This creates a reference implementation for the client team.
    """
    print(f"\n=== SERVER DECRYPTION REQUIREMENTS ===")
    
    # Use a fixed key and data for deterministic output
    fixed_key = b'\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10' + \
                b'\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20'
    
    test_content = b"Test file content for backup"
    
    print(f"Fixed AES Key: {binascii.hexlify(fixed_key).decode()}")
    print(f"Test Content: {test_content}")
    print(f"Content Length: {len(test_content)} bytes")
    
    # Show what client must produce
    encrypted_output = simulate_client_aes_encryption(test_content, fixed_key)
    
    print(f"\nClient must produce this encrypted output:")
    print(f"Hex: {binascii.hexlify(encrypted_output).decode()}")
    print(f"Length: {len(encrypted_output)} bytes")
    
    # Verify server can decrypt it
    print(f"\nVerifying server can decrypt...")
    decrypted = simulate_server_aes_decryption(encrypted_output, fixed_key)
    
    success = (decrypted == test_content)
    print(f"Server decryption result: {'✅ SUCCESS' if success else '❌ FAILED'}")
    
    if success:
        print(f"\n✅ REFERENCE IMPLEMENTATION CONFIRMED")
        print(f"If the client produces the above hex output for the given input,")
        print(f"the server will successfully decrypt it back to the original content.")
    
    return success

if __name__ == "__main__":
    print("Starting AES compatibility validation...")
    
    # Run full test suite
    all_passed = run_compatibility_tests()
    
    # Show reference implementation
    ref_success = demonstrate_exact_server_decryption()
    
    # Final status
    if all_passed and ref_success:
        sys.exit(0)  # Success
    else:
        sys.exit(1)  # Failure

File: test_aes_compatibility.cpp
// Test AES-256-CBC compatibility with Python server
// This test verifies that our C++ AES implementation produces exactly
// the same output as the Python server expects

#include "aes_crypto.h"
#include <iostream>
#include <vector>
#include <string>
#include <iomanip>
#include <cassert>

// Test vector: Known AES-256-CBC with zero IV
void testKnownVector() {
    std::cout << "\n=== TESTING KNOWN AES-256-CBC VECTOR ===" << std::endl;
    
    // Test key (32 bytes for AES-256) - matches Python test
    std::vector<uint8_t> testKey = {
        0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
        0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
        0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17,
        0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f
    };
    
    // Test plaintext (16 bytes - exactly one block) - matches Python test
    std::vector<uint8_t> plaintext = {
        'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o',
        'r', 'l', 'd', '!', ' ', 'T', 'e', 's'
    };
    
    // Expected encrypted result from Python server test
    std::vector<uint8_t> expected_encrypted = {
        0x83, 0x27, 0x30, 0x71, 0xcd, 0x50, 0x8b, 0x87,
        0x18, 0x09, 0xf6, 0x77, 0xe0, 0xc2, 0x86, 0x7b,
        0xa2, 0x15, 0x16, 0xcb, 0x02, 0xa1, 0xc5, 0xb1,
        0xf1, 0x63, 0xfd, 0xd7, 0xd3, 0x74, 0x0e, 0x8d,
        0x8e, 0x07, 0x7d, 0x8e, 0x14, 0x97, 0xcf, 0x61,
        0x22, 0x09, 0x12, 0x72, 0x9a, 0xbf, 0x95, 0x03
    };
    
    std::cout << "Key: " << AESCryptoUtils::bytesToHex(testKey) << std::endl;
    std::cout << "Plaintext: " << AESCryptoUtils::bytesToHex(plaintext) << std::endl;
    
    // Encrypt with our implementation
    AESCrypto crypto;
    if (!crypto.setAESKey(testKey)) {
        std::cout << "❌ FAILED: Could not set AES key" << std::endl;
        return;
    }
    
    try {
        std::vector<uint8_t> encrypted = crypto.encryptFileData(plaintext);
        std::cout << "Encrypted: " << AESCryptoUtils::bytesToHex(encrypted) << std::endl;
        std::cout << "Expected:  " << AESCryptoUtils::bytesToHex(expected_encrypted) << std::endl;
        
        // Test if our output matches the server's expected format
        bool format_match = (encrypted == expected_encrypted);
        std::cout << "Format compatibility: " << (format_match ? "✅ PASSED" : "❌ FAILED") << std::endl;
        
        // Decrypt to verify roundtrip
        std::vector<uint8_t> decrypted = crypto.decryptFileData(encrypted);
        std::cout << "Decrypted: " << AESCryptoUtils::bytesToHex(decrypted) << std::endl;
        
        bool success = (decrypted == plaintext);
        std::cout << "Roundtrip test: " << (success ? "✅ PASSED" : "❌ FAILED") << std::endl;
        
        if (success && format_match) {
            std::cout << "✅ AES-256-CBC with zero IV is working correctly and server-compatible!" << std::endl;
        } else if (success) {
            std::cout << "⚠️  AES-256-CBC works but format may not match server expectations" << std::endl;
        }
        
    } catch (const std::exception& e) {
        std::cout << "❌ FAILED: Exception during encryption/decryption: " << e.what() << std::endl;
    }
}

// Test with various data sizes to verify PKCS7 padding
void testPaddingBehavior() {
    std::cout << "\n=== TESTING PKCS7 PADDING BEHAVIOR ===" << std::endl;
    
    AESCrypto crypto;
    std::vector<uint8_t> testKey(32, 0x42); // Simple test key
    crypto.setAESKey(testKey);
    
    // Test different sizes to verify padding works correctly
    std::vector<size_t> testSizes = {1, 15, 16, 17, 31, 32, 33, 47, 48, 49, 63, 64, 65};
    
    for (size_t size : testSizes) {
        std::vector<uint8_t> testData(size, 0x55); // Fill with 0x55
        
        try {
            std::vector<uint8_t> encrypted = crypto.encryptFileData(testData);
            std::vector<uint8_t> decrypted = crypto.decryptFileData(encrypted);
            
            bool success = (decrypted == testData);
            std::cout << "Size " << std::setw(2) << size << " bytes: " 
                      << (success ? "✅ PASSED" : "❌ FAILED")
                      << " (encrypted: " << encrypted.size() << " bytes)" << std::endl;
            
            if (!success) {
                std::cout << "  Expected: " << testData.size() << " bytes" << std::endl;
                std::cout << "  Got:      " << decrypted.size() << " bytes" << std::endl;
            }
        } catch (const std::exception& e) {
            std::cout << "Size " << std::setw(2) << size << " bytes: ❌ FAILED (exception: " << e.what() << ")" << std::endl;
        }
    }
}

// Test server compatibility by simulating server-side decryption
void testServerCompatibility() {
    std::cout << "\n=== TESTING SERVER COMPATIBILITY ===" << std::endl;
    
    AESCrypto crypto;
    std::vector<uint8_t> testKey(32);
    
    // Generate a reproducible test key
    for (int i = 0; i < 32; i++) {
        testKey[i] = static_cast<uint8_t>((i * 7) % 256);
    }
    
    crypto.setAESKey(testKey);
    
    // Test data that might be in a real file
    std::string fileContent = "This is a test file that will be encrypted and sent to the server.\n"
                             "It contains multiple lines of text to verify that the encryption\n"
                             "works correctly with different content sizes and patterns.\n"
                             "The server should be able to decrypt this successfully.\n";
    
    std::vector<uint8_t> fileData(fileContent.begin(), fileContent.end());
    
    std::cout << "Original file size: " << fileData.size() << " bytes" << std::endl;
    
    try {
        // Encrypt the file data
        std::vector<uint8_t> encrypted = crypto.encryptFileData(fileData);
        std::cout << "Encrypted file size: " << encrypted.size() << " bytes" << std::endl;
        
        // Verify the encrypted size is properly padded
        if (encrypted.size() % 16 != 0) {
            std::cout << "❌ FAILED: Encrypted data is not block-aligned" << std::endl;
            return;
        }
        
        // Verify decryption works
        std::vector<uint8_t> decrypted = crypto.decryptFileData(encrypted);
        
        if (decrypted == fileData) {
            std::cout << "✅ PASSED: Server compatibility test successful!" << std::endl;
            std::cout << "   - Encrypted data is properly padded" << std::endl;
            std::cout << "   - Decryption restores original data exactly" << std::endl;
            std::cout << "   - Zero IV and PKCS7 padding are working correctly" << std::endl;
        } else {
            std::cout << "❌ FAILED: Decrypted data doesn't match original" << std::endl;
            std::cout << "   Original: " << fileData.size() << " bytes" << std::endl;
            std::cout << "   Decrypted: " << decrypted.size() << " bytes" << std::endl;
        }
        
    } catch (const std::exception& e) {
        std::cout << "❌ FAILED: Exception during server compatibility test: " << e.what() << std::endl;
    }
}

// Test that our implementation produces the same output format as expected by server
void testServerExpectedFormat() {
    std::cout << "\n=== TESTING SERVER EXPECTED FORMAT ===" << std::endl;
    
    AESCrypto crypto;
    
    // Use the exact same key format that would come from server
    std::vector<uint8_t> serverKey(32, 0x77); // Server-generated key
    crypto.setAESKey(serverKey);
    
    // Sample file content
    std::vector<uint8_t> fileContent = {'T', 'e', 's', 't', ' ', 'f', 'i', 'l', 'e'};
    
    try {
        std::vector<uint8_t> encrypted = crypto.encryptFileData(fileContent);
        
        std::cout << "Encryption format verification:" << std::endl;
        std::cout << "  - Input size: " << fileContent.size() << " bytes" << std::endl;
        std::cout << "  - Output size: " << encrypted.size() << " bytes" << std::endl;
        std::cout << "  - Block alignment: " << (encrypted.size() % 16 == 0 ? "✅ CORRECT" : "❌ WRONG") << std::endl;
        
        // The server expects:
        // 1. AES-256-CBC encryption
        // 2. Zero IV (16 bytes of zeros)
        // 3. PKCS7 padding
        // 4. The result should be decryptable with the same parameters
        
        // Simulate what the server does:
        // cipher_aes = AES.new(current_aes_key, AES.MODE_CBC, iv=b'\0' * 16)
        // decrypted_data = unpad(cipher_aes.decrypt(full_encrypted_data), AES.block_size)
        
        std::vector<uint8_t> decrypted = crypto.decryptFileData(encrypted);
        
        if (decrypted == fileContent) {
            std::cout << "✅ FORMAT VERIFICATION PASSED!" << std::endl;
            std::cout << "   The encrypted data format is compatible with server expectations" << std::endl;
        } else {
            std::cout << "❌ FORMAT VERIFICATION FAILED!" << std::endl;
        }
        
    } catch (const std::exception& e) {
        std::cout << "❌ FORMAT VERIFICATION FAILED: " << e.what() << std::endl;
    }
}

int main() {
    std::cout << "🔒 AES-256-CBC Server Compatibility Test Suite" << std::endl;
    std::cout << "===============================================" << std::endl;
    std::cout << "Testing AES implementation against Python server requirements:" << std::endl;
    std::cout << "- Algorithm: AES-256-CBC" << std::endl;
    std::cout << "- Key size: 32 bytes (256 bits)" << std::endl;
    std::cout << "- IV: Zero IV (16 bytes of zeros)" << std::endl;
    std::cout << "- Padding: PKCS7" << std::endl;
    std::cout << "- Block size: 16 bytes" << std::endl;
    
    testKnownVector();
    testPaddingBehavior();
    testServerCompatibility();
    testServerExpectedFormat();
    
    std::cout << "\n=== SUMMARY ===" << std::endl;
    std::cout << "If all tests passed, the AES implementation is ready for use with the simple client." << std::endl;
    std::cout << "The encrypted data should be compatible with the Python server's decryption process." << std::endl;
    
    return 0;
}